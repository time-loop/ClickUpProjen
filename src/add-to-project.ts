import { typescript, YamlFile } from 'projen';

export module addToProjectWorkflow {
  export const RENOVATE_GITHUB_USERNAME = 'cu-infra-svc-git';

  export const DEFAULT_RENOVATE_PR_LABEL = 'renovate';
  export const GITHUB_PROJECT_NUMBER = '3';

  const defaultWorkflow = {
    name: 'add-to-project',
    on: {
      pull_request: {
        types: ['opened', 'labeled'],
      },
    },
    // Prevent renovate running multiple times in parallel due to several subsequent trigger events occurring
    concurrency: {
      group: "${{ github.workflow }}-${{ github.event_name == 'workflow_dispatch' && github.event_name || '' }}",
    },
    jobs: {
      'add-to-project': {
        'runs-on': 'ubuntu-latest',
        permissions: {
          contents: 'read',
        },
        steps: [
          {
            name: 'Checkout',
            uses: 'actions/checkout@v3',
          },
          {
            name: 'Add issue to project',
            uses: 'actions/add-to-project@v0.4.1',
            with: {
              // github project url
              'project-url': `https://github.com/orgs/time-loop/projects/${GITHUB_PROJECT_NUMBER}`,
              'github-token': 'x-access-token:${{ secrets.PROJEN_GITHUB_TOKEN }}',
              labeled: DEFAULT_RENOVATE_PR_LABEL,
            },
          },
        ],
      },
    },
  };

  export function addAddToProjectWorkflowYml(project: typescript.TypeScriptProject, override?: any): void {
    new YamlFile(project, '.github/workflows/add-to-project.yml', { obj: { ...defaultWorkflow, ...override } });
  }
}
