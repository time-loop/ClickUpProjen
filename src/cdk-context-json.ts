import { awscdk, JsonPatch, SampleFile } from 'projen';

export module cdkContextJson {
  export interface InjectionOptions {
    /**
     * The account number where the role the GH actions will assume lives.
     * This is usually the same account running cdk-pipelines,
     * but can be any account which has been configured with
     *
     * ```ts
     * cdk bootstrap --trust-for-lookup $this_account_number ...
     * ```
     */
    readonly lookupAccountId: string;
    /**
     * Which region should GH Auth into?
     * @default 'us-west-2'
     */
    readonly awsRegion?: string;
    /**
     * How long should the requested role be valid.
     * @default 900 15min seems pretty generous.
     */
    readonly roleDurationSeconds?: number;
  }

  /**
   * Options to support cdk.context.json management
   */
  export interface Options {
    /**
     * @default - do not create stack SampleCode for OIDC role
     */
    readonly createOidcRoleStack?: boolean;
    /**
     * You must configure this if you want to self-mutation cdk.context.json.
     *
     * @default - do not modify the build workflow to authorize against AWS.
     */
    readonly injectionOptions?: InjectionOptions;
  }

  export function injectAwsAuthIntoBuild(project: awscdk.AwsCdkTypeScriptApp, options: InjectionOptions): void {
    const build = project.tryFindObjectFile('.github/workflows/build.yml');
    build?.addOverride('jobs.build.permissions', { 'id-token': 'write' });
    build?.patch(
      JsonPatch.add('/jobs/build/steps/0', {
        name: 'Configure AWS Credentials',
        uses: 'aws-actions/configure-aws-credentials@v2',
        with: {
          'aws-region': options.awsRegion ?? 'us-west-2',
          'role-to-assume': `arn:aws:iam::${options.lookupAccountId}:role/${project.name}-github-actions-role`,
          'role-duration-seconds': options.roleDurationSeconds ?? 900,
        },
      }),
    );
  }

  export function addOidcRoleStack(project: awscdk.AwsCdkTypeScriptApp): void {
    new SampleFile(project, `${project.srcdir}/github-actions-oidc-cdk-context-lookup-role.ts`, {
      contents: `import { core } from '@time-loop/cdk-library';
import { aws_iam, Stage } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { Namer } from 'multi-convention-namer';

export class GitHubActionsOIDCCdkContextLookupRole extends core.Stack {
  static asStage(scope: Construct, stageName: string, stageProps: core.StageProps): Stage {
    return new (class extends Stage {
      constructor() {
        super(scope, stageName, stageProps);
        new GitHubActionsOIDCCdkContextLookupRole(this, stageProps);
      }
    })();
  }
  constructor(scope: Construct, props: core.StackProps) {
    const projectName = '${project.name}';
    // NOTE: you should never deploy this anywhere but the cdkPipelines account
    // And the GitHubActionsOIDCPermissions class for cdk-diff should never be deployed
    // in the cdkPipelines account. Name collision is intentional here.
    let id = new Namer([...projectName.split('-'), 'github', 'actions']);
    super(scope, id, props);

    const githubActionsRoleName = id.addSuffix(['role']).kebab;
    const githubActionsRole = new aws_iam.Role(this, githubActionsRoleName, {
      roleName: githubActionsRoleName,
      assumedBy: new aws_iam.FederatedPrincipal(
        \`arn:aws:iam::\${this.account}:oidc-provider/token.actions.githubusercontent.com\`,
        {
          StringEquals: {
            'token.actions.githubusercontent.com:aud': 'sts.amazonaws.com',
          },
          StringLike: {
            'token.actions.githubusercontent.com:sub': \`repo:time-loop/\${projectName}:*\`,
          },
        },
        'sts:AssumeRoleWithWebIdentity',
      ),
    });

    const githubActionsPolicyName = id.addSuffix(['policy']).kebab;
    const githubActionsPolicy = new aws_iam.Policy(this, githubActionsPolicyName, {
      policyName: githubActionsPolicyName,
      statements: [
        new aws_iam.PolicyStatement({
          actions: [ 'sts:AssumeRole' ],
          resources: [ 'arn:aws:iam::*:role/cdk-*lookup-role*' ],
        }),
      ],
    });

    // Attach IAM policy to IAM role
    githubActionsPolicy.attachToRole(githubActionsRole);
  }
}
`,
    });
  }
}
