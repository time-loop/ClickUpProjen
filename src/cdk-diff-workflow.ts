import { typescript, YamlFile } from 'projen';
import { NodePackage } from 'projen/lib/javascript';

export module cdkDiffWorkflow {
  function createCdkDiffWorkflow(options: CDKDiffOptionsConfig) {
    const defaultWorkflow = {
      name: 'cdk-diff',
      on: {
        pull_request: {},
        workflow_dispatch: {},
      },
      permissions: {
        'id-token': 'write',
        contents: 'write',
        'pull-requests': 'write',
      },
      jobs: {
        diff: {
          'runs-on': 'ubuntu-latest',
          env: {
            CI: 'true',
            GITHUB_PR_NUMBER: '${{github.event.pull_request.number}}',
            REPO_NAME: '${{ github.event.repository.name }}',
          },
          steps: [
            {
              name: 'Checkout',
              uses: 'actions/checkout@v3',
              with: {
                ref: '${{ github.event.pull_request.head.ref }}',
                repository: '${{ github.event.pull_request.head.repo.full_name }}',
              },
            },
            {
              name: 'GitHub Packages authorization',
              run: 'cat > .npmrc <<EOF\n//npm.pkg.github.com/:_authToken=${{ secrets.ALL_PACKAGE_READ_TOKEN }}\n@time-loop:registry=https://npm.pkg.github.com/\nEOF',
            },
            {
              name: 'Setup Node.js',
              uses: 'actions/setup-node@v3',
              with: {
                'node-version': '14.x',
              },
            },
            {
              name: 'Install dependencies',
              run: 'yarn install --check-files',
            },
            {
              name: 'ls stacks in each env',
              run: './node_modules/.bin/cdk ls -l -j > cdk-ls.json\nQA_STACKS=$(jq -r \'[.[].id | select(contains("Qa")) ] | join(" ")\' cdk-ls.json)\nSTG_STACKS=$(jq -r \'[.[].id | select(contains("Staging")) ] | join(" ")\' cdk-ls.json)\nPROD_STACKS=$(jq -r \'[.[].id | select(contains("Prod")) ] | join(" ")\' cdk-ls.json)\necho "QA_STACKS=$QA_STACKS" >> $GITHUB_ENV\necho "STG_STACKS=$STG_STACKS" >> $GITHUB_ENV\necho "PROD_STACKS=$PROD_STACKS" >> $GITHUB_ENV\n',
            },
            {
              name: 'configure qa aws credentials',
              uses: 'aws-actions/configure-aws-credentials@v1',
              with: {
                'role-to-assume': options.oidcQaRoleArn,
                'role-duration-seconds': 900,
                'aws-region': 'us-west-2',
              },
            },
            {
              name: 'diff qa',
              run: "set -o pipefail\n./node_modules/.bin/cdk diff $QA_STACKS --progress=events --staging false -e --ignore-errors --version-reporting false &> >(tee cdk-qa.log)\n./node_modules/.bin/ts-node ./node_modules/@time-loop/cdk-log-parser/src/cdkLogParser.ts cdk-qa.log && echo 'QA_HAS_NO_DIFF=true' >> $GITHUB_ENV\n",
            },
            {
              name: 'configure staging aws credentials',
              uses: 'aws-actions/configure-aws-credentials@v1',
              with: {
                'role-to-assume': options.oidcStagingRoleArn,
                'role-duration-seconds': 900,
                'aws-region': 'us-west-2',
              },
            },
            {
              name: 'diff staging',
              run: "set -o pipefail\n./node_modules/.bin/cdk diff $STG_STACKS --progress=events --staging false -e --ignore-errors --version-reporting false &> >(tee cdk-staging.log)\n./node_modules/.bin/ts-node ./node_modules/@time-loop/cdk-log-parser/src/cdkLogParser.ts cdk-staging.log && echo 'STAGING_HAS_NO_DIFF=true' >> $GITHUB_ENV\n",
            },
            {
              name: 'configure prod aws credentials',
              uses: 'aws-actions/configure-aws-credentials@v1',
              with: {
                'role-to-assume': options.oidcProdRoleArn,
                'role-duration-seconds': 900,
                'aws-region': 'us-west-2',
              },
            },
            {
              name: 'diff prod',
              run: "set -o pipefail\n./node_modules/.bin/cdk diff $PROD_STACKS --progress=events --staging false -e --ignore-errors --version-reporting false &> >(tee cdk-prod.log)\n./node_modules/.bin/ts-node ./node_modules/@time-loop/cdk-log-parser/src/cdkLogParser.ts cdk-prod.log && echo 'PROD_HAS_NO_DIFF=true' >> $GITHUB_ENV\n",
            },
            {
              name: 'cdk-notify',
              run: "curl -L \"https://github.com/karlderkaefer/cdk-notifier/releases/download/v2.3.2/cdk-notifier_$(uname)_amd64.gz\" -o cdk-notifier.gz\ngunzip cdk-notifier.gz && chmod +x cdk-notifier && rm -rf cdk-notifier.gz\n./cdk-notifier --log-file ./cdk-qa.log --repo $REPO_NAME --owner $GITHUB_REPOSITORY_OWNER --token ${{ secrets.GITHUB_TOKEN }} --pull-request-id ${{ github.event.pull_request.number }} -t 'Qa Stacks'\n./cdk-notifier --log-file ./cdk-staging.log --repo $REPO_NAME --owner $GITHUB_REPOSITORY_OWNER --token ${{ secrets.GITHUB_TOKEN }} --pull-request-id ${{ github.event.pull_request.number }} -t 'Staging Stacks'\n./cdk-notifier --log-file ./cdk-prod.log --repo $REPO_NAME --owner $GITHUB_REPOSITORY_OWNER --token ${{ secrets.GITHUB_TOKEN }} --pull-request-id ${{ github.event.pull_request.number }} -t 'Production Stacks'\n",
            },
            {
              name: 'Apply qa has no diff label based on diff status',
              if: "env.QA_HAS_NO_DIFF == 'true'",
              uses: 'actions/github-script@v6',
              with: {
                'github-token': '${{ secrets.GITHUB_TOKEN }}',
                script:
                  'github.rest.issues.addLabels({\n  issue_number: context.issue.number,\n  owner: context.repo.owner,\n  repo: context.repo.repo,\n  labels: ["no-qa-changes"]\n})\n',
              },
            },
            {
              name: 'Apply staging has no diff label based on diff status',
              if: "env.STAGING_HAS_NO_DIFF == 'true'",
              uses: 'actions/github-script@v6',
              with: {
                'github-token': '${{ secrets.GITHUB_TOKEN }}',
                script:
                  'github.rest.issues.addLabels({\n  issue_number: context.issue.number,\n  owner: context.repo.owner,\n  repo: context.repo.repo,\n  labels: ["no-staging-changes"]\n})\n',
              },
            },
            {
              name: 'Apply prod has no diff label based on diff status',
              if: "env.PROD_HAS_NO_DIFF == 'true'",
              uses: 'actions/github-script@v6',
              with: {
                'github-token': '${{ secrets.GITHUB_TOKEN }}',
                script:
                  'github.rest.issues.addLabels({\n  issue_number: context.issue.number,\n  owner: context.repo.owner,\n  repo: context.repo.repo,\n  labels: ["no-prod-changes"]\n})\n',
              },
            },
            {
              name: 'Should auto merge',
              id: 'should-auto-merge',
              if: "env.QA_HAS_NO_DIFF == 'true' && env.STAGING_HAS_NO_DIFF == 'true' && env.PROD_HAS_NO_DIFF == 'true' && github.event.pull_request.user.login == 'cu-infra-svc-git'",
              run: 'echo "auto-merge=true" >> $GITHUB_OUTPUT',
            },
            {
              name: 'Enable automerge',
              uses: 'daneden/enable-automerge-action@v1.0.2',
              if: 'steps.should-auto-merge.outputs.auto-merge',
              with: {
                'github-token': '${{ secrets.PROJEN_GITHUB_TOKEN }}',
                'merge-method': 'SQUASH',
                'allowed-author': 'cu-infra-svc-git',
              },
            },
            {
              name: 'Auto approve',
              uses: 'hmarr/auto-approve-action@v2.2.1',
              if: 'steps.should-auto-merge.outputs.auto-merge',
              with: {
                'github-token': '${{ secrets.GITHUB_TOKEN }}',
              },
            },
          ],
        },
      },
    };

    return defaultWorkflow;
  }

  export interface CDKDiffOptionsConfig {
    /**
     * Name of the QA OIDC role name which contains neccesasry IAM policies to run the CDK diff
     *
     * Example arn: `arn:aws:iam::123123123123:role/squad-github-actions-permissions-squad-cdk-github-actions-role`
     */
    readonly oidcQaRoleArn: string;

    /**
     * Name of the Staging OIDC role name which contains neccesasry IAM policies to run the CDK diff
     *
     * Example arn: `arn:aws:iam::123123123123:role/squad-github-actions-permissions-squad-cdk-github-actions-role`
     */
    readonly oidcStagingRoleArn: string;

    /**
     * Name of the Prod OIDC role name which contains neccesasry IAM policies to run the CDK diff
     *
     * Example arn: `arn:aws:iam::123123123123:role/squad-github-actions-permissions-squad-cdk-github-actions-role`
     */
    readonly oidcProdRoleArn: string;
  }

  export function getCDKDiffOptions(options?: CDKDiffOptionsConfig) {
    return options;
  }

  export function addCdkDiffWorkflowYml(
    project: typescript.TypeScriptProject,
    options: CDKDiffOptionsConfig,
    override?: any,
  ): void {
    new YamlFile(project, '.github/workflows/cdk-diff.yml', {
      obj: { ...createCdkDiffWorkflow(options), ...override },
    });
  }

  export function AddCdkLogParserDependency(pkg: NodePackage) {
    pkg.addDevDeps('@time-loop/cdk-log-parser@0.0.0');
  }
}
