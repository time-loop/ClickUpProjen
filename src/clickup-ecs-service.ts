import * as fs from 'fs';
import * as path from 'path';
import { Component, SampleFile } from 'projen';
import merge from 'ts-deepmerge';
import { clickupTs } from './clickup-ts';
import { renovateWorkflow } from './renovate-workflow';
import { ecsServiceCDWorkflow } from './workflows/ecs-service-cd';
import { ecsServiceCIWorkflow } from './workflows/ecs-service-ci';
import { ecsServiceDeployWorkflow } from './workflows/ecs-service-deploy';

export module clickupEcsService {
  export const defaults: Omit<ClickUpTypeScriptEcsServiceProjectOptions, 'name' | 'serviceName'> = {
    ...clickupTs.defaults,
    workflowNodeVersion: '16',
  };

  export interface ClickUpTypeScriptEcsServiceProjectOptions extends clickupTs.ClickUpTypeScriptProjectOptions {
    readonly deployWorkflowOptions?: ecsServiceDeployWorkflow.EcsServiceDeployOptionsConfig;
    readonly serviceName: string;
    readonly serviceHealthcheckPath?: string;
  }

  export class ClickUpTypeScriptEcsServiceProject extends clickupTs.ClickUpTypeScriptProject {
    readonly workflowNodeVersion: string;
    readonly serviceName: string;
    readonly serviceHealthcheckPath: string;

    constructor(options: ClickUpTypeScriptEcsServiceProjectOptions) {
      const serviceHealthcheckPath = options.serviceHealthcheckPath ?? `${options.serviceName}/health`;
      let mergedOptions = merge(
        defaults,
        {
          deps: clickupTs.deps,
          deployWorkflowOptions: {
            deployEnvs: [ecsServiceDeployWorkflow.ClickUpHarnessEnv.QA],
            serviceHealthcheckPath,
          },
        },
        options,
        {
          // Disable projen's built-in docgen class
          docgen: undefined,
          name: clickupTs.normalizeName(options.name),
          renovatebotOptions: renovateWorkflow.getRenovateOptions(options.renovateOptionsConfig),
        },
      );
      super(mergedOptions);
      this.workflowNodeVersion = mergedOptions.workflowNodeVersion!;
      this.serviceName = options.serviceName;
      this.serviceHealthcheckPath = serviceHealthcheckPath;

      new AppSampleCode(this);

      ecsServiceCIWorkflow.createContinuousIntegrationWorkflow(this, {
        ...mergedOptions,
        nodeVersion: this.workflowNodeVersion,
      });

      ecsServiceCDWorkflow.createContinuousDeliveryWorkflow(this, {
        ...mergedOptions,
        ...mergedOptions.deployWorkflowOptions!,
        nodeVersion: this.workflowNodeVersion,
      });
    }
  }

  class AppSampleCode extends Component {
    constructor(project: ClickUpTypeScriptEcsServiceProject) {
      super(project);

      new SampleFile(project, 'Dockerfile', {
        contents: this.renderSample('Dockerfile.sample'),
      });

      new SampleFile(project, 'scripts/docker-healthcheck.sh', {
        contents: this.renderSample('docker-healthcheck.sh.sample', {
          '<%SERVICE_HEALTHCHECK_PATH%>': project.serviceHealthcheckPath,
        }),
      });

      new SampleFile(project, 'scripts/docker-build.sh', {
        contents: this.renderSample('docker-build.sh.sample', {
          '<%SERVICE_NAME%>': project.serviceName,
        }),
      });
    }

    private getSample(fileName: string): string {
      return fs.readFileSync(path.join(__dirname, 'samples', fileName), 'utf-8');
    }

    private renderSample(sampleFileName: string, vars?: { [toReplace: string]: string }): string {
      let contents = this.getSample(sampleFileName);

      if (vars) {
        const toReplace = Object.keys(vars);
        toReplace.map((key) => {
          contents = contents.replace(new RegExp(key, 'g'), vars[key]);
        });
      }
      return contents;
    }
  }
}
