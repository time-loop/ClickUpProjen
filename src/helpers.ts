/**
 * Strips out ^ and ~ characters from package names in favor of specific
 * versions.
 * @param theDeps A list of dependencies (and possibly their versions)
 * @returns A list of dependencies pinned to a version
 */
export function getPinnedDeps(theDeps: string[]): string[] {
  return theDeps.map((d) => {
    return d.replace(new RegExp(/~|\^/), '');
  });
}
