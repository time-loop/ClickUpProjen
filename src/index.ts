export * from './clickup-cdk';
export * from './clickup-ts';
export * from './datadog';
export * from './slack-alert';
export * from './renovate-workflow';
export * from './cdk-diff-workflow';
export * from './add-to-project';
