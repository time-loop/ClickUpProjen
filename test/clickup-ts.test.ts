import { Testing } from 'projen';
import { clickupTs } from '../src';
import { getPinnedDeps } from '../src/helpers';

describe('ClickUpTypeScriptProject', () => {
  describe('defaults', () => {
    const p = new clickupTs.ClickUpTypeScriptProject({
      name: '@time-loop/test',
      defaultReleaseBranch: 'main',
    });
    const synth = Testing.synth(p);
    // console.log(Object.getOwnPropertyNames(synth));
    ['codecov.yml', 'package.json', '.github/workflows/upgrade-main.yml', '.mergify.yml'].forEach((f) => {
      test(f, () => {
        expect(synth[f]).toMatchSnapshot();
      });
    });
    test('prettier is enabled', () => {
      expect(p.prettier).toBeTruthy();
    });
    test('jest is enabled', () => {
      expect(p.jest).toBeTruthy();
    });
    // TODO: soooo many more tests need to be written here.
  });
  describe('name', () => {
    let envCache = process.env;

    beforeEach(() => {
      process.env = envCache;
    });

    test('should add prefix when missing', () => {
      jest.spyOn(console, 'log');
      const p = new clickupTs.ClickUpTypeScriptProject({
        name: 'missing-prefix',
        defaultReleaseBranch: 'main',
      });
      expect(p.name).toBe('@time-loop/missing-prefix');
    });

    test('should respect GITHUB_OWNER', () => {
      jest.spyOn(console, 'log');
      process.env.GITHUB_OWNER = 'fizzle';
      const p = new clickupTs.ClickUpTypeScriptProject({
        name: 'missing-prefix',
        defaultReleaseBranch: 'main',
      });
      expect(p.name).toBe('@fizzle/missing-prefix');
    });
  });
  describe('docgen', () => {
    const projenDocgenTaskName = 'docgen';
    const clickupDocgenTaskName = 'typedocDocgen';
    const commonOpts: clickupTs.ClickUpTypeScriptProjectOptions = {
      name: 'random-prefix',
      defaultReleaseBranch: 'main',
    };

    test('succeeds with defaults', () => {
      new clickupTs.ClickUpTypeScriptProject(commonOpts);
    });

    test('includes markdown plugin by default', () => {
      const project = new clickupTs.ClickUpTypeScriptProject(commonOpts);
      expect(project.deps.getDependency('typedoc-plugin-markdown')).toBeDefined();
    });

    test('excludes markdown plugin when html attr is set', () => {
      const project = new clickupTs.ClickUpTypeScriptProject({ ...commonOpts, docgenOptions: { html: true } });
      const getDep = () => {
        project.deps.getDependency('typedoc-plugin-markdown');
      };
      expect(getDep).toThrowError();
    });

    test('includes typedocDocgen task by default', () => {
      const project = new clickupTs.ClickUpTypeScriptProject(commonOpts);
      expect(project.tasks.tryFind(clickupDocgenTaskName)).toBeDefined();
      expect(project.tasks.tryFind(projenDocgenTaskName)).toBeUndefined();
    });

    test('contains no docgen task when docgen is false', () => {
      const project = new clickupTs.ClickUpTypeScriptProject({ ...commonOpts, docgen: false });
      [clickupDocgenTaskName, projenDocgenTaskName].forEach((task) =>
        expect(project.tasks.tryFind(task)).toBeUndefined(),
      );
    });

    test('creates typedoc config JSON file', () => {
      const project = new clickupTs.ClickUpTypeScriptProject(commonOpts);
      const matchedFiles = project.files.filter((file) => file.path === 'typedoc.json');
      expect(matchedFiles).toHaveLength(1);
    });

    test('respects configFilePath attr when set', () => {
      const testOverride = 'test.json';
      const project = new clickupTs.ClickUpTypeScriptProject({
        ...commonOpts,
        docgenOptions: { configFilePath: testOverride },
      });
      const matchedFiles = project.files.filter((file) => file.path === testOverride);
      expect(matchedFiles).toHaveLength(1);
    });
  });
  describe('peerDeps', () => {
    const commonOpts: clickupTs.ClickUpTypeScriptProjectOptions = {
      name: 'random-prefix',
      defaultReleaseBranch: 'main',
      peerDeps: ['@time-loop/cdk-ecs-fargate@^1.0.3', 'ts-deepmerge@2.5.3'],
    };
    const project = new clickupTs.ClickUpTypeScriptProject(commonOpts);
    const synth = Testing.synth(project);
    const pkgJson = synth['package.json'];

    it('is set correctly', () => {
      // Need to stringify, then parse to make this accessible
      const peerDeps: Record<string, string> = JSON.parse(JSON.stringify(pkgJson.peerDependencies));
      expect(peerDeps).toBeDefined();
      const actualPeerDeps = Object.entries(peerDeps).map((kvPair) => {
        return kvPair.join('@');
      });
      commonOpts.peerDeps!.forEach((dep) => {
        expect(actualPeerDeps.includes(dep)).toBeTruthy();
      });
    });
    it('adds to devDeps correctly', () => {
      // Need to stringify, then parse to make this accessible
      const devDeps: Record<string, string> = JSON.parse(JSON.stringify(pkgJson.devDependencies));
      expect(devDeps).toBeDefined();
      const actualDevDeps = Object.entries(devDeps).map((kvPair) => {
        return kvPair.join('@');
      });
      const expectedDevDeps = getPinnedDeps(commonOpts.peerDeps!);
      expectedDevDeps.forEach((dep) => {
        expect(actualDevDeps.includes(dep)).toBeTruthy();
      });
    });
  });
});

describe('normalizeName', () => {
  const envCache = process.env;

  beforeEach(() => {
    process.env = envCache;
    delete process.env.GITHUB_OWNER;
  });

  test('should silently pass when prefix is present', () => {
    const logSpy = jest.spyOn(console, 'log');
    const name = clickupTs.normalizeName('@time-loop/has-prefix');
    expect(name).toBe('@time-loop/has-prefix');
    expect(logSpy).not.toHaveBeenCalled();
  });

  test('should add prefix and complain when missing', () => {
    const logSpy = jest.spyOn(console, 'log');
    const name = clickupTs.normalizeName('missing-prefix');
    expect(name).toBe('@time-loop/missing-prefix');
    expect(logSpy).toHaveBeenCalledWith(
      'Adding mandatory prefix @time-loop/ to name. New name: @time-loop/missing-prefix',
    );
  });

  test('should respect GITHUB_OWNER', () => {
    const logSpy = jest.spyOn(console, 'log'); // silence console.log
    process.env.GITHUB_OWNER = 'fizzle';
    const name = clickupTs.normalizeName('missing-prefix');
    expect(name).toBe('@fizzle/missing-prefix');
    expect(logSpy).toHaveBeenCalled();
  });
});
