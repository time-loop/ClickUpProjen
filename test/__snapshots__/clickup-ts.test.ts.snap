// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClickUpTypeScriptProject defaults .github/workflows/upgrade-main.yml 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

name: upgrade-main
on:
  workflow_dispatch: {}
  schedule:
    - cron: 0 0 * * 1
jobs:
  upgrade:
    name: Upgrade
    runs-on: ubuntu-latest
    permissions:
      contents: read
    outputs:
      patch_created: \${{ steps.create_patch.outputs.patch_created }}
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          ref: main
      - name: GitHub Packages authorization
        run: |-
          cat > .npmrc <<EOF
          //npm.pkg.github.com/:_authToken=\${{ secrets.ALL_PACKAGE_READ_TOKEN }}
          @time-loop:registry=https://npm.pkg.github.com/
          EOF
      - name: Make cdk-ecr-deployment sane
        run: export FORCE_PREBUILT_LAMBDA=1
      - name: Setup Node.js
        uses: actions/setup-node@v2.2.0
        with:
          node-version: 14.17.0
      - name: Install dependencies
        run: yarn install --check-files --frozen-lockfile
      - name: Upgrade dependencies
        run: npx projen upgrade
      - id: create_patch
        name: Find mutations
        run: |-
          git add .
          git diff --staged --patch --exit-code > .repo.patch || echo \\"::set-output name=patch_created::true\\"
      - if: steps.create_patch.outputs.patch_created
        name: Upload patch
        uses: actions/upload-artifact@v2
        with:
          name: .repo.patch
          path: .repo.patch
  pr:
    name: Create Pull Request
    needs: upgrade
    runs-on: ubuntu-latest
    permissions:
      contents: write
      pull-requests: write
    if: \${{ needs.upgrade.outputs.patch_created }}
    steps:
      - name: Checkout
        uses: actions/checkout@v3
        with:
          token: \${{ secrets.PROJEN_GITHUB_TOKEN }}
          ref: main
      - name: Download patch
        uses: actions/download-artifact@v3
        with:
          name: .repo.patch
          path: \${{ runner.temp }}
      - name: Apply patch
        run: '[ -s \${{ runner.temp }}/.repo.patch ] && git apply \${{ runner.temp }}/.repo.patch || echo \\"Empty patch. Skipping.\\"'
      - name: Set git identity
        run: |-
          git config user.name \\"github-actions\\"
          git config user.email \\"github-actions@github.com\\"
      - name: Create Pull Request
        id: create-pr
        uses: peter-evans/create-pull-request@v3
        with:
          token: \${{ secrets.PROJEN_GITHUB_TOKEN }}
          commit-message: |-
            chore(deps): upgrade dependencies

            Upgrades project dependencies. See details in [workflow run].

            [Workflow Run]: https://github.com/\${{ github.repository }}/actions/runs/\${{ github.run_id }}

            ------

            *Automatically created by projen via the \\"upgrade-main\\" workflow*
          branch: github-actions/upgrade-main
          title: \\"chore(deps): upgrade dependencies\\"
          body: |-
            Upgrades project dependencies. See details in [workflow run].

            [Workflow Run]: https://github.com/\${{ github.repository }}/actions/runs/\${{ github.run_id }}

            ------

            *Automatically created by projen via the \\"upgrade-main\\" workflow*
          author: github-actions <github-actions@github.com>
          committer: github-actions <github-actions@github.com>
          signoff: true
"
`;

exports[`ClickUpTypeScriptProject defaults .mergify.yml 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

queue_rules:
  - name: default
    conditions:
      - \\"#approved-reviews-by>=1\\"
      - -label~=(do-not-merge)
      - status-success=build
pull_request_rules:
  - name: Automatic merge on approval and successful build
    actions:
      delete_head_branch: {}
      queue:
        method: squash
        name: default
        commit_message_template: |-
          {{ title }} (#{{ number }})

          {{ body }}
    conditions:
      - \\"#approved-reviews-by>=1\\"
      - -label~=(do-not-merge)
      - status-success=build
"
`;

exports[`ClickUpTypeScriptProject defaults codecov.yml 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

coverage:
  precision: 2
  round: down
  status:
    project:
      default:
        target: auto
        threshold: 10%
        paths:
          - src
        if_ci_failed: error
        only_pulls: true
    patch:
      default:
        base: auto
        target: auto
        threshold: 10%
        paths:
          - src
        if_ci_failed: error
        only_pulls: true
parsers:
  gcov:
    branch_detection:
      conditional: yes
      loop: yes
      method: no
      macro: no
comment:
  layout: reach,diff,flags,files,footer
  behavior: default
  require_changes: no
"
`;

exports[`ClickUpTypeScriptProject defaults package.json 1`] = `
Object {
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "author": Object {
    "name": "ClickUp",
    "organization": true,
  },
  "dependencies": Object {
    "@time-loop/clickup-projen": "*",
  },
  "devDependencies": Object {
    "@types/jest": "^27",
    "@types/node": "^14",
    "@typescript-eslint/eslint-plugin": "^5",
    "@typescript-eslint/parser": "^5",
    "esbuild": "*",
    "eslint": "^8",
    "eslint-config-prettier": "*",
    "eslint-import-resolver-node": "*",
    "eslint-import-resolver-typescript": "*",
    "eslint-plugin-import": "*",
    "eslint-plugin-prettier": "*",
    "jest": "^27",
    "jest-junit": "^13",
    "jsii-release": "*",
    "json-schema": "*",
    "npm-check-updates": "^15",
    "prettier": "*",
    "projen": "*",
    "standard-version": "^9",
    "ts-jest": "^27",
    "ts-node": "^10",
    "typescript": "*",
  },
  "engines": Object {
    "node": ">= 14.17.0",
  },
  "jest": Object {
    "clearMocks": true,
    "collectCoverage": true,
    "collectCoverageFrom": Array [
      "src/**/*.ts",
    ],
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": Array [
      "/node_modules/",
    ],
    "coverageReporters": Array [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text",
    ],
    "globals": Object {
      "ts-jest": Object {
        "tsconfig": "tsconfig.dev.json",
      },
    },
    "preset": "ts-jest",
    "reporters": Array [
      "default",
      Array [
        "jest-junit",
        Object {
          "outputDirectory": "test-reports",
        },
      ],
    ],
    "testMatch": Array [
      "<rootDir>/src/**/__tests__/**/*.ts?(x)",
      "<rootDir>/(test|src)/**/*(*.)@(spec|test).ts?(x)",
    ],
    "testPathIgnorePatterns": Array [
      "/node_modules/",
    ],
    "watchPathIgnorePatterns": Array [
      "/node_modules/",
    ],
  },
  "license": "UNLICENSED",
  "main": "lib/index.js",
  "name": "@time-loop/test",
  "publishConfig": Object {
    "registry": "https://npm.pkg.github.com/",
  },
  "scripts": Object {
    "build": "npx projen build",
    "bump": "npx projen bump",
    "clobber": "npx projen clobber",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "eject": "npx projen eject",
    "eslint": "npx projen eslint",
    "package": "npx projen package",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "projen": "npx projen",
    "release": "npx projen release",
    "test": "npx projen test",
    "test:update": "npx projen test:update",
    "test:watch": "npx projen test:watch",
    "unbump": "npx projen unbump",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch",
  },
  "types": "lib/index.d.ts",
  "version": "0.0.0",
}
`;
